src/utils/BufferReader.hpp

```diff
@@ -2,26 +2,24 @@
 #ifndef BUFFER_READER_HPP
 #define BUFFER_READER_HPP
 
-#include <utils/Buffer.hpp>
+#include "utils/IBuffer.hpp"
 
 /* ************************************************************************** */
 class BufferReader
 {
 public:
-  void init(Buffer* buffer);
+  void init(IBuffer* buffer);
```

Comment:
I don't see a reason why `BufferReader` needs a two-phase initialization. With a constructor all the asserts in the source file can go away.
[NR.5: Don’t use two-phase initialization](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#rnr-two-phase-init)
```suggestion
  explicit BufferReader(IBuffer& buffer);
```